<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: calc_rank_dist_ave Summarize pairwise differences.</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for calc_rank_dist_ave {primerTree}"><tr><td>calc_rank_dist_ave {primerTree}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>calc_rank_dist_ave
Summarize
pairwise differences.</h2>

<h3>Description</h3>

<p>calc_rank_dist_ave
Summarize
pairwise differences.
</p>


<h3>Usage</h3>

<pre>
calc_rank_dist_ave(x, ranks = common_ranks, ...)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>x</code></td>
<td>
<p>a primerTree object</p>
</td></tr>
<tr valign="top"><td><code>ranks</code></td>
<td>
<p>ranks to show unique counts for, defaults to the common ranks</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>
<p>Ignored options</p>
</td></tr>
</table>


<h3>Details</h3>

<p>The purpose of this function is to calculate the average number
of nucleotide differences between species within a given taxonomic
level.
</p>
<p>There are several key assumptions and calculations made in this
function.
</p>
<p>First, the function randomly selects one sequence from each species
in the primerTree results. This is to keep any one species (e.g.
human, cow, etc.) with many hits from skewing the results.
</p>
<p>Second, for each taxonomic level tested, the function divides the
sequences by each taxon at that level and calculates the mean
number of nucleotide differences within that taxa, then returns the
mean of those values.
</p>
<p>Third, when calculating the average distance, any taxa for which
there is only one species is omitted, as the number of nucleotide
differences will always be 0.
</p>


<h3>Value</h3>

<p>returns a data frame of results
</p>


<h3>Examples</h3>

<pre>
## Not run: 
calc_rank_dist_ave(mammals_16S)

calc_rank_dist_ave(bryophytes_trnL)

# Note that the differences between the results from these two primers
# the mean nucleotide differences is much higher for the mammal primers
# than the byrophyte primers. This suggests that the mammal primers have
# better resolution to distinguish individual species.

## End(Not run)
</pre>

<hr /><div style="text-align: center;">[Package <em>primerTree</em> version 1.0.3 <a href="00Index.html">Index</a>]</div>
</body></html>
